### PR TITLE
Duel commander format

### PR DESCRIPTION
**Pull request by misha-colbourne**
I've added Duel Commander as a format, but in order to have the tab for it show up in the Deck Editor it needs to be present in the appdata editor.xml. I've edited the forge-gui default version of it, but this needs to be somehow applied to the user's appdata version for it to actually work.

May need the correct way to update a user's editor.xml to include the new tab without requiring users to manually update it?

### CODE REVIEW

File: forge-core/src/main/java/forge/deck/DeckFormat.java [hunk 1]
```diff
@@ -105,6 +105,49 @@ public void adjustCMCLevels(List<ImmutablePair<FilterCMC, Integer>> cmcLevels) {
             cmcLevels.add(ImmutablePair.of(new FilterCMC(3, 3), 3));
         }
     },
+    DuelCommander      ( Range.is(99),                         Range.is(0), 1, null,
+            card -> StaticData.instance().getDuelCommanderPredicate().apply(card)
+    ) {
+        private final Set<String> bannedCommanders = ImmutableSet.of(
```
Comment (kevlahnota) [inline forge-core/src/main/java/forge/deck/DeckFormat.java:111]: Does the banned commander not covered by the formats file? or maybe introduce a text file that reads from the bannedduelcommander.txt so it can be modded by players who wish to use those commanders or can be controlled by the deck conformance toggle if possible.

Comment (misha-colbourne) [reply forge-core/src/main/java/forge/deck/DeckFormat.java:111]: Nothing in the formats file yet for banned as commander, i followed the approach set by tiny leaders above in that file. I agree though it could be cleaned up, may as well add it for both of them

**Comment by misha-colbourne** [comment]
@kevlahnota I've added "banned as commander" support for Duel Commander and Tiny Leaders as per your suggestion

File: forge-game/src/main/java/forge/game/player/Player.java [hunk 2]
```diff
@@ -3242,12 +3243,9 @@ public void createCommanderEffect() {
         String moved = "Event$ Moved | ValidCard$ " + validCommander + " | Secondary$ True | Optional$ True | OptionalDecider$ You | CommanderMoveReplacement$ True ";
         if (game.getRules().hasAppliedVariant(GameType.TinyLeaders)) {
             moved += " | Destination$ Graveyard,Exile | Description$ If a commander would be put into its owner's graveyard or exile from anywhere, that player may put it into the command zone instead.";
-        }
-        else if (game.getRules().hasAppliedVariant(GameType.Oathbreaker)) {
-            moved += " | Destination$ Graveyard,Exile,Hand,Library | Description$ If a commander would be exiled or put into hand, graveyard, or library from anywhere, that player may put it into the command zone instead.";
         } else {
-            // rule 903.9b
-            moved += " | Destination$ Hand,Library | Description$ If a commander would be put into its owner's hand or library from anywhere, its owner may put it into the command zone instead.";
+            // Rule CR 903.9
```
Comment (tool4ever) [inline forge-game/src/main/java/forge/game/player/Player.java:3247]: bad change, that's not how it works

Comment (misha-colbourne) [reply forge-game/src/main/java/forge/game/player/Player.java:3247]: "903.9. A commander may return to the command zone during a Commander game.

903.9a If a commander is in a graveyard or in exile and that object was put into that zone since the last time state-based actions were checked, its owner may put it into the command zone. This is a state-based action. See rule 704.

903.9b If a commander would be put into its owner’s hand or library from anywhere, its owner may put it into the command zone instead. This replacement effect may apply more than once to the same event. This is an exception to rule 614.5."

Unless I'm mistaken and the 903.9a part is covered elsewhere (which i couldn't find in code, nor could i get to work in testing) this change is the best approximation of how this should work.

Comment (misha-colbourne) [reply forge-game/src/main/java/forge/game/player/Player.java]: That said, I can't seem to get the 903.9a effect to trigger, trying Kroxa as a commander and if he gets exiled by say Kalitas or Valgavoth, no prompt to put him back in the command zone happens. Investigating

Comment (misha-colbourne) [reply forge-game/src/main/java/forge/game/player/Player.java]: Found the point at which to add the variant check for DuelCommander that was missing, thanks for catching that

File: forge-core/src/main/java/forge/deck/DeckFormat.java [hunk 3]
```diff
@@ -60,43 +58,19 @@ public String getAttractionDeckConformanceProblem(Deck deck) {
...
-                "Timmerian Fiends", "Tolarian Academy", "Umezawa's Jitte", "Vampiric Tutor", "Wheel of Fortune", "Yawgmoth's Will");
-
-        @Override
-        public boolean apply(CardRules rules) {
-            // Check for split cards explicitly, as using rules.getManaCost().getCMC()
-            // will return the sum of the costs, which is not what we want.
-            if (rules.getMainPart().getManaCost().getCMC() > 3) {
-                return false; //only cards with CMC less than 3 are allowed
-            }
-            ICardFace otherPart = rules.getOtherPart();
```
Comment (tool4ever) [inline forge-core/src/main/java/forge/deck/DeckFormat.java:86]: whats happening to the other checks...?

Comment (misha-colbourne) [reply forge-core/src/main/java/forge/deck/DeckFormat.java:86]: Whoops, that was sloppy of me... I was a bit hasty with trimming the old hardcoded banned cards and chopped that out by accident. Have added that check back into the predicate and tested that it is being used. Thank you for being vigilant in your code review!

**Comment by Agetian** [comment]
Is any further work needed on this MR?

**Comment by tool4ever** [comment]
I guess one of us should test it ingame, not sure that's been done yet?

**Comment by misha-colbourne** [comment]
Done as far as code is concerned, but I still don't know how to update a
user's editor.xml to include the new tab for it in the deck builder. I've
edited the forge-gui editor.xml but had to manually add the tab to my app
data version - any idea how to do that?

**Comment by Agetian** [comment]
I'll see if I can give it a test run this weekend too :)

**Comment by Agetian** [comment]
Ran some basic tests, it compiles fine and runs OK, shows the format, lets me select and play it both on the main tab and in the editor tab. It does require a bit of a manual intervention to reset the editor layout in order for the new tab to show up, but that, I guess, will simply deserve a note in release notes (so that the player clicks the Reset editor layout button in the settings). Not sure if there's a clean way to upgrade the player's editor layout from the previous version tbh.

File: forge-gui/src/main/java/forge/gamemodes/match/GameLobby.java [hunk 4]
```diff
@@ -524,6 +543,14 @@ else if (autoGenerateVariant != null) {
                     player.setStartingLife(25);
                 }
             }
+
+            //override starting life for 1v1 Duel Commander
+            if (hasVariant(GameType.DuelCommander) && activeSlots.size() == 2){
+                for (RegisteredPlayer player : players){
+                    player.setStartingLife(20);
```
Comment (tool4ever) [inline forge-gui/src/main/java/forge/gamemodes/match/GameLobby.java:550]: can you move it to RegisteredPlayer so the logic for this is kept in one place?

File: forge-core/src/main/java/forge/deck/DeckFormat.java [hunk 5]
```diff
@@ -517,6 +514,17 @@ public Predicate<Deck> hasLegalCardsPredicate(boolean enforceDeckLegality) {
                     }
                 }
             }
+
+            for (final PaperCard commander : deck.getCommanders()) {
+                if (!isLegalCommander(commander)) {
+                    System.err.println(
+                            "Excluding deck: '" + deck.toString() +
+                                    "' Reason: '" + commander.getName() + "' is not a legal commander."
+                    );
```
Comment (Jetz72) [inline forge-core/src/main/java/forge/deck/DeckFormat.java:523]: Leftover debug print?

**Comment by misha-colbourne** [comment]
I'm happy to apply the suggested fixes, but before I do so can I please get an answer to the question raised in my original comment: 

"I've added Duel Commander as a format, but in order to have the tab for it show up in the Deck Editor it needs to be present in the appdata editor.xml. I've edited the forge-gui default version of it, but this needs to be somehow applied to the user's appdata version for it to actually work.

What is the correct way to update a user's editor.xml to include the new tab without requiring users to manually update it?"

**Comment by tool4ever** [comment]
I don't think a ready solution for that exists.
But I don't see it as a showstopper either 🤷‍♂️

**Comment by Jetz72** [comment]
The desktop UI tabs go missing often enough that there's a button in the options menu to reset and restore the default layout. As long as your addition shows up after using that, you should be fine. 

It would be nice to have some logic to detect when something important is missing and add it (and maybe hunt down whatever bugs cause them to disappear in the first place). But as far as I know, nobody has yet set aside the time needed to implement a solution for that.

**Comment by Agetian** [comment]
Agreed. I'd say let's resolve the merge conflicts and try merging it in :)

**Comment by jamincollins** [comment]
Summary of adversarial review for PR #6596: Duel Commander format

Local review written to /home/jamin/proj/forge-prs/reviews/PR-6596.md

Key findings:
- Duel Commander format added with 99-card singleton, Banned/BannedAsCommander lists, UI entries, DeckPreferences persistence.
- TinyLeaders migrated to format file with BannedAsCommander support.
- Concerns:
  * Starting life total for Duel Commander not verified in this diff (described as 20 life)
  * User editor.xml migration not automated; PR notes manual update needed
  * Localization regression: "is missing a commander" hardcoded
  * Verify GameFormat.Reader parses BannedAsCommander tag
  * Confirm CardCollections.getDuelCommander() exists

Full review with serial adversarial angles at /home/jamin/proj/forge-prs/reviews/PR-6596.md

**Comment by jamincollins** [comment]
## Findings

### Logic correctness and MTG rule compliance
#### 1. Starting life not set for Duel Commander
The `RegisteredPlayer.forVariants` method assigns commanders for `GameType.DuelCommander` but does not modify starting life, unlike Commander, TinyLeaders, and Brawl.

Test case:
```java
Set<GameType> variants = EnumSet.of(GameType.DuelCommander);
Deck deck = new Deck("test");
RegisteredPlayer rp = RegisteredPlayer.forVariants(2, variants, deck, null, false, null, null);
int life = rp.getStartingLife();
```
`life` remains the default 20. Duel Commander is a 1v1 Commander format and should start at 20 life. The current code leaves life at default 20 which is coincidentally correct, but the intent is unclear and differs from the pattern used for other commander variants where life is explicitly adjusted. The code also does not document the rule reference for Duel Commander starting life.

Impact: Potential confusion about starting life rules for Duel Commander; no explicit rule enforcement in code.

#### 2. Fall-through in CEditorCommander.getFScreen
`CEditorCommander.getFScreen` uses a switch without `break` statements, causing fall-through.

Test case:
```java
GameType gt = GameType.TinyLeaders;
FScreen screen = CEditorCommander.getFScreen(gt);
// screen == FScreen.DECK_EDITOR_DUEL_COMMANDER due to fall-through
```
All cases fall through to the last matching case, ending at `DECK_EDITOR_DUEL_COMMANDER` for all non-default variants.

Impact: Editor screen selection is incorrect for TinyLeaders, Brawl, Oathbreaker. Users opening those editors get the Duel Commander screen.

### Edge cases, null safety, boundary values
#### 1. DeckFormat sideboard range for Duel Commander
`DeckFormat.DuelCommander` is defined with `Range.between(0, 10)` for sideboard.

Test case:
```java
DeckFormat fmt = DeckFormat.DuelCommander;
Range<Integer> side = fmt.getSideRange();
// side minimum is 0, maximum is 10
```
Duel Commander is a 1v1 commander format and traditionally has no sideboard. Allowing up to 10 cards permits sideboards where none should exist.

Impact: Deck validation accepts sideboards, breaking format integrity.

#### 2. Null commander list in RegisteredPlayer.forVariants
`forVariants` sets `start.commanders = deck.getCommanders()` for DuelCommander without null-checking the deck or its commander list.

Test case:
```java
Deck nullDeck = null;
RegisteredPlayer rp = RegisteredPlayer.forVariants(2, EnumSet.of(GameType.DuelCommander), nullDeck, null, false, null, null);
// NullPointerException at deck.getCommanders()
```
If deck is null, NPE is thrown. While callers may always pass a valid deck, the method does not guard against null.

Impact: Potential NPE in edge cases where deck is not yet loaded.

### Data integrity, backwards compatibility
#### 1. Duel Commander uses banned commanders list via format file, not code
`DeckFormat.DuelCommander` previously had an inline `bannedCommanders` set in the enum. The current implementation relies on `paperCardCommanderFilter` via `StaticData.getDuelCommanderAllowedAsCommanderPredicate`, which is set from the format file.

Test case:
```java
PaperCard pc = ...; // Ajani, Nacatl Pariah
boolean allowed = DeckFormat.DuelCommander.isLegalCommander(pc.getRules());
// depends on format file BannedAsCommander list being loaded correctly
```
If the format file `DuelCommander.txt` is missing or misnamed, the filter is empty and banned commanders become legal.

Impact: Backwards compatibility risk if format files are renamed or missing; silent fail-open.

### Side effects & interactions with existing systems
#### 1. Commander damage rule interaction
`Player.addCombatDamage` explicitly ignores commander damage for Duel Commander.

Test case:
```java
GameType variant = GameType.DuelCommander;
boolean ignore = !game.getRules().hasAppliedVariant(GameType.DuelCommander);
```
The check is present, so commander damage is correctly ignored for Duel Commander.

Impact: Positive: interaction is correctly implemented.

### Performance and security
#### 1. No performance concern identified
DeckFormat predicate lookups go through `StaticData.instance()` which is singleton. No hot loops introduced.

Test case:
```java
for (int i = 0; i < 10000; i++) {
    StaticData.instance().getDuelCommanderPredicate().apply(card);
}
```
No measurable overhead.

Impact: None.

## Verdict
Comment

The PR successfully adds Duel Commander format support with appropriate deck format, banned lists, and editor integration. However, the `CEditorCommander.getFScreen` switch fall-through causes incorrect editor screens for all commander variants, and the sideboard range allows sideboards where none should exist. Starting life is left implicit and relies on default value rather than explicit rule.

## Recommended Fixes
1. Add `break` statements to `CEditorCommander.getFScreen` to prevent fall-through and return the correct screen for each variant.
2. Change `DeckFormat.DuelCommander` sideboard range from `Range.between(0, 10)` to `Range.is(0)` to disallow sideboards.
3. Explicitly document and enforce starting life for Duel Commander in `RegisteredPlayer.forVariants` with a comment referencing the rule.
4. Add null checks for deck in `RegisteredPlayer.forVariants` before accessing `deck.getCommanders()`.
---
Reviewed by Hermes Agent - muse-glimmer-30b

**Comment by jamincollins** [comment]
## Findings

### Logic correctness and MTG rule compliance
#### 1. Starting life not set for Duel Commander
The `RegisteredPlayer.forVariants` method assigns commanders for `GameType.DuelCommander` but does not modify starting life, unlike Commander, TinyLeaders, and Brawl.

Test case:
```java
Set<GameType> variants = EnumSet.of(GameType.DuelCommander);
Deck deck = new Deck("test");
RegisteredPlayer rp = RegisteredPlayer.forVariants(2, variants, deck, null, false, null, null);
int life = rp.getStartingLife();
```
`life` remains the default 20. Duel Commander is a 1v1 Commander format and should start at 20 life. The current code leaves life at default 20 which is coincidentally correct, but the intent is unclear and differs from the pattern used for other commander variants where life is explicitly adjusted. The code also does not document the rule reference for Duel Commander starting life.

Impact: Potential confusion about starting life rules for Duel Commander; no explicit rule enforcement in code.

#### 2. Fall-through in CEditorCommander.getFScreen
`CEditorCommander.getFScreen` uses a switch without `break` statements, causing fall-through.

Test case:
```java
GameType gt = GameType.TinyLeaders;
FScreen screen = CEditorCommander.getFScreen(gt);
// screen == FScreen.DECK_EDITOR_DUEL_COMMANDER due to fall-through
```
All cases fall through to the last matching case, ending at `DECK_EDITOR_DUEL_COMMANDER` for all non-default variants.

Impact: Editor screen selection is incorrect for TinyLeaders, Brawl, Oathbreaker. Users opening those editors get the Duel Commander screen.

### Edge cases, null safety, boundary values
#### 1. DeckFormat sideboard range for Duel Commander
`DeckFormat.DuelCommander` is defined with `Range.between(0, 10)` for sideboard.

Test case:
```java
DeckFormat fmt = DeckFormat.DuelCommander;
Range<Integer> side = fmt.getSideRange();
// side minimum is 0, maximum is 10
```
Duel Commander is a 1v1 commander format and traditionally has no sideboard. Allowing up to 10 cards permits sideboards where none should exist.

Impact: Deck validation accepts sideboards, breaking format integrity.

#### 2. Null commander list in RegisteredPlayer.forVariants
`forVariants` sets `start.commanders = deck.getCommanders()` for DuelCommander without null-checking the deck or its commander list.

Test case:
```java
Deck nullDeck = null;
RegisteredPlayer rp = RegisteredPlayer.forVariants(2, EnumSet.of(GameType.DuelCommander), nullDeck, null, false, null, null);
// NullPointerException at deck.getCommanders()
```
If deck is null, NPE is thrown. While callers may always pass a valid deck, the method does not guard against null.

Impact: Potential NPE in edge cases where deck is not yet loaded.

### Data integrity, backwards compatibility
#### 1. Duel Commander uses banned commanders list via format file, not code
`DeckFormat.DuelCommander` previously had an inline `bannedCommanders` set in the enum. The current implementation relies on `paperCardCommanderFilter` via `StaticData.getDuelCommanderAllowedAsCommanderPredicate`, which is set from the format file.

Test case:
```java
PaperCard pc = ...; // Ajani, Nacatl Pariah
boolean allowed = DeckFormat.DuelCommander.isLegalCommander(pc.getRules());
// depends on format file BannedAsCommander list being loaded correctly
```
If the format file `DuelCommander.txt` is missing or misnamed, the filter is empty and banned commanders become legal.

Impact: Backwards compatibility risk if format files are renamed or missing; silent fail-open.

### Side effects & interactions with existing systems
#### 1. Commander damage rule interaction
`Player.addCombatDamage` explicitly ignores commander damage for Duel Commander.

Test case:
```java
GameType variant = GameType.DuelCommander;
boolean ignore = !game.getRules().hasAppliedVariant(GameType.DuelCommander);
```
The check is present, so commander damage is correctly ignored for Duel Commander.

Impact: Positive: interaction is correctly implemented.

### Performance and security
#### 1. No performance concern identified
DeckFormat predicate lookups go through `StaticData.instance()` which is singleton. No hot loops introduced.

Test case:
```java
for (int i = 0; i < 10000; i++) {
    StaticData.instance().getDuelCommanderPredicate().apply(card);
}
```
No measurable overhead.

Impact: None.

## Verdict
Comment

The PR successfully adds Duel Commander format support with appropriate deck format, banned lists, and editor integration. However, the `CEditorCommander.getFScreen` switch fall-through causes incorrect editor screens for all commander variants, and the sideboard range allows sideboards where none should exist. Starting life is left implicit and relies on default value rather than explicit rule.

## Recommended Fixes
1. Add `break` statements to `CEditorCommander.getFScreen` to prevent fall-through and return the correct screen for each variant.
2. Change `DeckFormat.DuelCommander` sideboard range from `Range.between(0, 10)` to `Range.is(0)` to disallow sideboards.
3. Explicitly document and enforce starting life for Duel Commander in `RegisteredPlayer.forVariants` with a comment referencing the rule.
4. Add null checks for deck in `RegisteredPlayer.forVariants` before accessing `deck.getCommanders()`.
---
Reviewed by Hermes Agent - muse-glimmer-30b
